[LLVM][Instrumentation] Add numerical sanitizer

alexander-shaposhnikov opened the pull request:
This PR introduces the numerical sanitizer originally proposed by Clement Courbet on https://reviews.llvm.org/D97854
(https://arxiv.org/abs/2102.12782).

The main additions include:
- Migration to LLVM opaque pointers
- Migration to various updated APIs
- Extended coverage for LLVM instructions/intrinsics, that enabled us to have a green run of tests again

The tool is still very experimental, the coverage (e.g. for intrinsics / library functions) is incomplete.

Test plan:
1. ninja check-all
2. ninja check-nsan

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+  // Do not create shadow stacks for intrinsics/known lib funcs.
+  if (Function *Fn = CI.getCalledFunction()) {
+    LibFunc LFunc;
+    if (Fn->getIntrinsicID() || TLI.getLibFunc(*Fn, LFunc))
```
check isIntrinsic or getIntrinsicID == Intrinsic::not_intrinsic

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+  if (addrPointsToConstantData(Load.getPointerOperand())) {
+    // No need to look into the shadow memory, the value is a constant. Just
+    // convert from FT to 2FT.
+    return Builder.CreateCast(Instruction::FPExt, &Load, ExtendedVT);
```
CreateFPExt

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+  //     fpext double  %1 to x86_fp80  ->  fpext float    s(%1) to double
+
+  // See (A) above.
+  Value *const Source =
```
east const throughout?

alexander-shaposhnikov replied:
Removed

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+                                                   IRBuilder<> &Builder) {
+  // We cannot look inside inline asm, just expand the result again.
+  if (Call.isInlineAsm()) {
+    return Builder.CreateCast(Instruction::FPExt, &Call, ExtendedVT);
```
CreateFPExt

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+    Builder.SetCurrentDebugLocation(Extract->getDebugLoc());
+    // FIXME: We could make aggregate object first class citizens. For now we
+    // just extend the extracted value.
+    return Builder.CreateCast(Instruction::FPExt, Extract, ExtendedVT);
```
CreateFPExt

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+        continue; // Not an FT value.
+      if (Map.hasShadow(Op))
+        continue; // Shadow is already available.
+      assert(isa<Instruction>(Op) &&
+             "non-instructions should already have shadows");
+      assert(!isa<PHINode>(Op) && "phi nodes should aready have shadows");
```
Just use cast below instead of assert(isa) + dyn_cast

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+             "non-instructions should already have shadows");
+      assert(!isa<PHINode>(Op) && "phi nodes should aready have shadows");
+      MissingShadow = true;
+      DfsStack.push_back(dyn_cast<Instruction>(Op));
```
```suggestion
      DfsStack.push_back(cast<Instruction>(Op));
```

alexander-shaposhnikov replied:
(left the assert in place cause it makes debugging slightly easier, replaced dyn_cast with cast)

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+    // appropriate size.
+    Type *BitcastTy = nullptr; // The FT type to bitcast to.
+    if (ConstantInt *CInt = dyn_cast<ConstantInt>(C)) {
+      switch (CInt->getType()->getScalarSizeInBits()) {
```
Better to switch over fltSemantics?  I think there's a helper for this already

alexander-shaposhnikov replied:
need a bit more details how fltSemantics would help here

arsenm replied:
bfloat/half and ppc_f128 / fp128 have the same bit widths but different float types. I assume we have a helper for this already somewhere

alexander-shaposhnikov replied:
So this is for debugging purposes only (added nsan-propagate-non-ft-const-stores-as-ft to control it),
as you've correctly pointed out the mapping is ambiguous, but luckily we don't really need it.

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2261 @@
+  initialize(*F.getParent());
+  SmallVector<Instruction *, 8> AllLoadsAndStores;
+  SmallVector<Instruction *, 8> LocalLoadsAndStores;
+  if (!F.hasFnAttribute(Attribute::SanitizeNumericalStability))
+    return false;
```
Can do this before all the initialize stuff?

alexander-shaposhnikov replied:
Done

vitalybuka commented:
Can we split the patch into smaller pieces?
e.g. clang, transform, runtime ?

This one has transform, comments, so we can keep it for transform.

vitalybuka commented:
clang changes LGTM, but they are missing testst

alexander-shaposhnikov commented:
@vitalybuka , @arsenm - thanks a lot, yeah, sure, I can split the patch.
P.S. will update/address comments soon.

alexander-shaposhnikov committed:
Add numerical sanitizer

alexander-shaposhnikov commented:
@arsenm , @vitalybuka : I've split out the changes to clang and compiler-rt from this PR.
Addressed comments (except one).

arsenm commented on llvm/test/Instrumentation/NumericalStabilitySanitizer/cfg.ll:
```diff
@@ -0,0 +1,113 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -mtriple=x86_64-linux-gnu -passes=nsan -nsan-shadow-type-mapping=dqq -S %s | FileCheck %s
```
This should go in test/Instrumentation/NumericalStabilitySanitizer/X86, or needs a REQUIRES: x86-registered-target

alexander-shaposhnikov replied:
Removed -mtriple=x86_64-linux-gnu

arsenm commented on llvm/test/Instrumentation/NumericalStabilitySanitizer/invoke.ll:
```diff
@@ -0,0 +1,148 @@
+  br label %exit
+
+land:
+  %res = landingpad { i8*, i32 } cleanup
```
Needs to use opaque pointers

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/test/Instrumentation/NumericalStabilitySanitizer/basic.ll:
```diff
@@ -0,0 +1,931 @@
+; CHECK-NEXT:    ret void
+;
+entry:
+  %b = load float, float* @float_const
```
opaque pointers

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+
+// If `FT` corresponds to a primitive FTValueType, return it.
+static std::optional<FTValueType> ftValueTypeFromType(Type *FT) {
+  if (FT->isFloatTy())
```
This should probably handle vectors, and at least half?

alexander-shaposhnikov replied:
Half-precision support is currently missing, but I'd say it's relatively easy to implement.
To avoid growing this patch further - maybe add this as a follow-up ? 
There are a number of missing features, I can file GitHub issues to track individual pieces of work.

P.S. This helper only deals with primitive types (non-vector).
(see e.g. MappingConfig::getExtendedFPType)

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+      // Check that the shadow type size is at most kShadowScale times the
+      // application type size, so that shadow memory compoutations are valid.
+      if (ShadowTypeSize > kShadowScale * AppTypeSize) {
+        errs() << "Invalid nsan mapping f" << AppTypeSize << "->f"
```
Can't spam errs in a pass. Use DiagnosticInfo?

alexander-shaposhnikov replied:
```
llvm-project/llvm/lib/Transforms$ grep -r -n "errs" ./* | wc -l
118
```
I'm aware of DiagnosticInfo, but I don't see anything appropriate on
https://llvm.org/doxygen/DiagnosticInfo_8h.html, it looks like the existing passes use errs() for similar purposes.

arsenm replied:
Same as above, this is not a common usage. Most of those look like they are in LLVM_DEBUG or dump

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+  const DataLayout &DL = M.getDataLayout();
+  Context = &M.getContext();
+  IntptrTy = DL.getIntPtrType(*Context);
+  Type *PtrTy = PointerType::getUnqual(*Context);
```
Don't use getUnqual, there's always a context you can pull the address space from

alexander-shaposhnikov replied:
so this is essentially only used by the initialization code (that populates NsanGetShadowPtrForStore) (10 lines below), so it has to be the default address space (can't be extracted from something else)

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+
+  Function *F = FCmp.getFunction();
+  if (CheckFunctionsFilter &&
+      !(F && CheckFunctionsFilter->match(F->getName()))) {
```
deMorgan this

alexander-shaposhnikov replied:
Removed the check for F / simplified

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+    // NOTE: Right now because of https://bugs.llvm.org/show_bug.cgi?id=45399
+    // for f128 we need to use Make_X86FP80_X86FP80 (go to a lower precision and
+    // come back).
+    {"llvm.sqrt.f32", Intrinsic::sqrt, Make_Double_Double},
```
You shouldn't need to maintain a mangled intrinsic name map. These names also aren't guaranteed, the suffix doesn't actually matter

alexander-shaposhnikov replied:
Yeah, I agree, this solution is not great. In its current state, it primarily functions as a way to create a static table (type, intrinsic, factory).  We  probably should replace this mechanism entirely, either by generating the table with TableGen or by eliminating the use of mangled names and passing type information differently.
I've tried the latter, but it's very tedious. I agree with the point, but honestly would prefer to fix it later.

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+  }
+}
+
+// Moves fast math flags from the function to individual instructions, and
+// removes the attribute from the function.
+// FIXME: Make this controllable with a flag.
+static void moveFastMathFlags(Function &F,
+                              std::vector<Instruction *> &Instructions) {
```
This isn't the place for this (if any such place should exist in the first place)

alexander-shaposhnikov replied:
this function exists to prevent flags from affecting shadow instructions (they are inserted into the same function) (this would be undesirable and actually incorrect).

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2263 @@
+      const int NumElements =
+          cast<VectorType>(CDV->getType())->getElementCount().getFixedValue();
+      switch (CDV->getType()->getScalarSizeInBits()) {
+      case 32:
```
Switch over fltSemantics? There are at least 2 size collisions already

alexander-shaposhnikov replied:
commented above, this whole thing that figures out which Bitcast to use is only necessary for debugging purposes.
I've added a flag ClPropagateNonFTConstStoresAsFT /* off by default */ to enable this branch,
but otherwise it can be removed.

arsenm commented on llvm/test/Instrumentation/NumericalStabilitySanitizer/basic.ll:
```diff
@@ -0,0 +1,931 @@
+  ret void
+}
+
+define void @constantload_add_store_double(double* %dst) sanitize_numericalstability {
```
opaque pointers

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+                                cl::desc("Check floating-point return values"),
+                                cl::Hidden);
+
+static const char *const kNsanModuleCtorName = "nsan.module_ctor";
+static const char *const kNsanInitName = "__nsan_init";
```
StringLiteral?

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+  case PPC128ShadowConfig::kNsanTypeId:
+    return std::make_unique<PPC128ShadowConfig>();
+  }
+  errs() << "nsan: invalid shadow type id'" << TypeId << "'\n";
```
Can't just print to errs

alexander-shaposhnikov replied:
llvm-project/llvm/lib/Transforms$ grep -r -n "errs" ./* | wc -l
118
I don't see anything appropriate on
https://llvm.org/doxygen/DiagnosticInfo_8h.html, it looks like the existing passes use errs() for similar purposes.

arsenm replied:
Existing instances of bad practice aren't a justification to add new instances of bad practices. Most of those contexts do not look like this type of use. If you want lazy error handling, I don't see why not just use report_fatal_error.

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+static FTValueType semanticsToFTValueType(const fltSemantics &Sem) {
+  if (&Sem == &APFloat::IEEEsingle()) {
+    return kFloat;
+  } else if (&Sem == &APFloat::IEEEdouble()) {
```
No else after return

alexander-shaposhnikov replied:
/* removed this code entirely */

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+// not an `enum class`.
+enum FTValueType { kFloat, kDouble, kLongDouble, kNumValueTypes };
+
+static FTValueType semanticsToFTValueType(const fltSemantics &Sem) {
```
Can make this a true switch with SemanticsToEnum

alexander-shaposhnikov replied:
Removed this conversion completely (and MappingConfig::bySemantics too), 
I think it was very confusing / was only used once.
So the main "entry point" for getting a shadow type is MappingConfig::getExtendedFPType.

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+  } else if (&Sem == &APFloat::x87DoubleExtended()) {
+    return kLongDouble;
+  }
+  llvm_unreachable("semantics are not one of the handled types");
```
Would be better to try to handle this more gracefully and just ignore it? We seem to be continuously adding new FP types

alexander-shaposhnikov replied:
/* removed this code entirely */

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+      return;
+  }
+
+  const DataLayout &DL =
+      CI.getParent()->getParent()->getParent()->getDataLayout();
```
Should just make DL a class member, or at least use getFunction to shrink the getParent chain

alexander-shaposhnikov replied:
switched to using getModule()

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+      CI.getParent()->getParent()->getParent()->getDataLayout();
+  // Set the shadow stack tag.
+  Builder.CreateStore(getCalleeAddress(CI, Builder), NsanShadowArgsTag);
+  unsigned ShadowArgsOffsetBytes = 0;
```
Maintain as TypeSize?

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+  if (Ty->isVectorTy()) {
+    auto *VecTy = cast<VectorType>(Ty);
+    Value *CheckResult = nullptr;
+    for (int I = 0, E = VecTy->getElementCount().getFixedValue(); I < E; ++I) {
```
Should avoid crashing on scalable vectors

alexander-shaposhnikov replied:
Done (we currently skip scalable vectors early in MappingConfig + added a test + FIXME)

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+                        ? Builder.CreateOr(CheckResult, ComponentCheckResult)
+                        : ComponentCheckResult;
+    }
+    assert(CheckResult && "struct with no FT element");
```
An empty struct is valid IR, shouldn't be an assert

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+  if (Instruction *Inst = dyn_cast<Instruction>(V)) {
+    Function *F = Inst->getFunction();
+    if (CheckFunctionsFilter &&
+        !(F && CheckFunctionsFilter->match(F->getName()))) {
```
Shouldn't need to worry about orphaned functions?

alexander-shaposhnikov replied:
Done.

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+      CheckResult, ConstantInt::get(Builder.getInt32Ty(), kResumeFromValue));
+  return Builder.CreateSelect(
+      ICmpEQ,
+      Builder.CreateCast(Instruction::FPExt, V,
```
Should use CreateFPExt

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+        Instruction::FPExt,
+        FCmpBuilder.CreateCast(Instruction::FPTrunc, ShadowLHS, LHS->getType()),
+        Ty);
+    ShadowRHS = FCmpBuilder.CreateCast(
+        Instruction::FPExt,
+        FCmpBuilder.CreateCast(Instruction::FPTrunc, ShadowRHS, RHS->getType()),
+        Ty);
```
CreateFPExt + CreateFPTrunc

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+  if (Function *Fn = Call.getCalledFunction()) {
+    // We're calling a statically known function.
+    return Builder.CreatePtrToInt(Fn, IntptrTy);
+  } else {
```
No else after return

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2265 @@
+
+// Returns a value with the address of the callee.
+Value *
+NumericalStabilitySanitizer::getCalleeAddress(CallBase &Call,
```
I don't see the point of this function, you can just store the pointer type directly without routing through ptrtoint

alexander-shaposhnikov replied:
Done

arsenm reviewed (commented):
There are a number of ways this is going to break on strictfp, but fixing that's probably a separate project for later

alexander-shaposhnikov commented:
@arsenm - many thanks for the review, will update the PR soon (finally got a successful run on Windows yesterday)

MaskRay commented:
(I've asked a few folks who study accuracy in floating-point arithmetic. I hope they can take a look as well.)

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2255 @@
+class ShadowTypeConfig {
+public:
+  static std::unique_ptr<ShadowTypeConfig> fromNsanTypeId(char TypeId);
+  // The floating-point semantics of the shadow type.
+  virtual const fltSemantics &semantics() const = 0;
+
+  // The LLVM Type corresponding to the shadow type.
+  virtual Type *getType(LLVMContext &Context) const = 0;
```
These 2 virtual are redundant, you can implement each one in terms of the other. From the Type you can directly query the semantics, and you can get the Type from Type::getFloatingPointTy.

The class also seems more complicated than necessary, you could replace it with a switch over the type to getNsanTypeId?

alexander-shaposhnikov replied:
Removed semantics(),
I think this hierarchy was set up this way in the original patch, I didn't change that.
Perhaps, having different configs (for different shadow types) here creates an extension point (e.g. for logging / debugging) (but yeah, one could organize things differently).

arsenm commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -819,6 +819,8 @@ static uint64_t getAttrKindEncoding(Attribute::AttrKind Kind) {
     return bitc::ATTR_KIND_SANITIZE_THREAD;
   case Attribute::SanitizeMemory:
     return bitc::ATTR_KIND_SANITIZE_MEMORY;
+  case Attribute::SanitizeNumericalStability:
+    return bitc::ATTR_KIND_SANITIZE_NUMERICAL_STABILITY;
```
Should add a bitcode compatibility test

alexander-shaposhnikov replied:
Done

jcranmer-intel commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+    assert(V);
+    assert(Shadow);
+    const bool Inserted = Map.emplace(V, Shadow).second;
+#ifdef LLVM_ENABLE_DUMP
```
I'd do `#if defined(LLVM_ENABLE_DUMP) && !defined(NDEBUG)` here, given that's essentially just providing more information on an assertion failure.

alexander-shaposhnikov replied:
Done

jcranmer-intel reviewed (commented):
Not comprehensive look yet:

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+// `dqq`.
+static cl::opt<std::string> ClShadowMapping(
+    "nsan-shadow-type-mapping", cl::init("dqq"),
+    cl::desc("One shadow type id for each of `float`, `double`, `long double`. "
```
The current schema is close to the source code, it is (the whole tool) kind of ~equivalent (not quite, but at least ~conceptually) to what one could do at the source level (use higher precision types).

alexander-shaposhnikov commented:
@arsenm , @vitalybuka , @jcranmer-intel - thanks a lot for the review, 
I've addressed most of the comments but added a few inline replies - would like to try to proceed here if possible

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+      return Configs[*VT]->getType(*Context);
+    if (FT->isVectorTy()) {
+      auto *VecTy = cast<VectorType>(FT);
+      // FIXME: add support for scalable vector types.
+      if (VecTy->isScalableTy())
+        return nullptr;
```
For this function, the code will work as-is for scalable since you use getElementCount

alexander-shaposhnikov replied:
so this needs to return `nullptr` cause the callers check it and skip / use fallback paths if no shadow type is available.
For scalable vectors specifically, there are a few places where they are currently not supported and we have asserts that we should not encounter them there yet (the asserts will trigger if we remove this).

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+    assert(V);
+    assert(Shadow);
+    const bool Inserted = Map.emplace(V, Shadow).second;
+#if defined(LLVM_ENABLE_DUMP) && !defined(NDEBUG)
```
LLVM_ENABLE_DUMP is for specific dump functions only, not enabling unconditional debug printing. This block should be wrapped in LLVM_DEBUG

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+#if defined(LLVM_ENABLE_DUMP) && !defined(NDEBUG)
+    if (!Inserted) {
+      if (const auto *const I = dyn_cast<Instruction>(V))
+        I->getParent()->getParent()->dump();
```
This is far too much context, just print the parent function name?

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+    }
+#endif
+    assert(Inserted && "duplicate shadow");
+    (void)Inserted;
```
[[maybe_unused]]?

alexander-shaposhnikov replied:
Done

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+  // Returns the shadow value for a given value. Asserts that the value has
+  // a shadow value. Lazily creates shadows for constant values.
+  Value *getShadow(Value *V) const {
+    assert(V);
+    if (Constant *C = dyn_cast<Constant>(V))
```
the assert is redundant with the dyn_cast

alexander-shaposhnikov replied:
Done

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+  case PPC128ShadowConfig::kNsanTypeId:
+    return std::make_unique<PPC128ShadowConfig>();
+  }
+  report_fatal_error("nsan: invalid shadow type id '" + Twine(TypeId) + "'");
```
there was a comment here about `errs()` - I've switched to report_fatal_error - so that one is done.

alexander-shaposhnikov commented:
Addressed comments

vitalybuka commented on llvm/include/llvm/IR/Attributes.td:
```diff
@@ -285,6 +285,9 @@ def SanitizeHWAddress : EnumAttr<"sanitize_hwaddress", [FnAttr]>;
 def SanitizeMemTag : EnumAttr<"sanitize_memtag", [FnAttr]>;
 
+/// NumericalStabilitySanitizer is on.
+def SanitizeNumericalStability : EnumAttr<"sanitize_numericalstability", [FnAttr]>;
```
Not sure what is the rule here, why not `sanitize_numerical_stability` ?

alexander-shaposhnikov replied:
will do (already switched on https://github.com/llvm/llvm-project/pull/93783)

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+} // end anonymous namespace
+
+PreservedAnalyses
+NumericalStabilitySanitizerPass::run(Function &F,
```
what is the point splitting to Fn and Module pass?

I'd recommend to move instrumentation into `NumericalStabilitySanitizerPass::run(Module &M, ModuleAnalysisManager &MAM)`

It's simplified some stuff for us in other sanitizers.

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+  }
+
+  MappingConfig *const Config;
+  std::unordered_map<Value *, Value *> Map;
```
Why not `DenseMap`?

vitalybuka replied:
and `Map` is not very descriptive

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+    llvm_unreachable("unimplemented");
+  }
+
+  MappingConfig *const Config;
```
if `Config` can't be null, please make it `&`

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+  bool sanitizeFunction(Function &F, const TargetLibraryInfo &TLI);
+
+private:
+  void initialize(Module &M);
```
`NumericalStabilitySanitizer(Module &M)`

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+  MappingConfig Config;
+  LLVMContext *Context = nullptr;
+  IntegerType *IntptrTy = nullptr;
+  FunctionCallee NsanGetShadowPtrForStore[FTValueType::kNumValueTypes];
```
`= {}` all these fields?

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+  if (GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(Addr))
+    Addr = GEP->getPointerOperand();
+
+  if (GlobalVariable *GV = dyn_cast<GlobalVariable>(Addr)) {
```
remove `{}`

alexander-shaposhnikov replied:
Done

alexander-shaposhnikov commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+static constexpr StringLiteral kNsanInitName("__nsan_init");
+
+// The following values must be kept in sync with the runtime.
+static constexpr const int kShadowScale = 2;
```
typo

alexander-shaposhnikov replied:
Done

MaskRay commented on llvm/test/Instrumentation/NumericalStabilitySanitizer/non_float_store.ll:
```diff
@@ -0,0 +1,20 @@
+}
+
+attributes #0 = { nounwind readonly uwtable sanitize_numericalstability "correctly-rounded-divide-sqrt-fp-math"="false" "denormal-fp-math"="preserve-sign,preserve-sign" "denormal-fp-math-f32"="ieee,ieee" "disable-tail-calls"="false" "frame-pointer"="none" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-infs-fp-math"="true" "no-jump-tables"="false" "no-nans-fp-math"="true" "no-signed ...
+
```
delete unneeded trailing blank lines

ditto elsewhere

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2233 @@
+// Internal part of emitCheck(). Returns a value that indicates whether
+// computation should continue with the shadow or resume by re-fextending the
+// value.
+enum ContinuationType { // Keep in sync with runtime.
```
enum class and remove `k` ?

vitalybuka replied:
same for the rest

alexander-shaposhnikov replied:
converted to enum class, the rest are a bit different (either enclosed by a class - for such enums llvm usually doesn't use enum class or used as an index (in an array), in this case we need the implicit conversion to int)

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+static MemoryExtents getMemoryExtentsOrDie(Type *FT) {
+  if (const auto VT = ftValueTypeFromType(FT))
+    return {*VT, 1};
+  if (FT->isVectorTy()) {
+    auto *VecTy = cast<VectorType>(FT);
```
fold into dyn_cast

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+  //     fptrunc fp128    %1 to x86_fp80  ->  fptrunc fp128     %1    to double
+
+  // See (A) above.
+  Value *const Source =
+      ExtendedSourceTy ? Map.getShadow(OrigSource) : OrigSource;
+  Type *const SourceTy = ExtendedSourceTy ? ExtendedSourceTy : OrigSourceTy;
```
no east const

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+  if (const char *Name = KnownIntrinsic::get(LFunc))
+    return Name;
+
+  errs() << "FIXME: LibFunc: " << TLI.getName(LFunc) << "\n";
```
wrap in LLVM_DEBUG?

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+  Value *IntrinsicCall = Builder.CreateIntrinsic(WidenedId, ArgTys, Args);
+  return WidenedFnTy->getReturnType() == ExtendedVT
+             ? IntrinsicCall
+             : Builder.CreateCast(Instruction::FPExt, IntrinsicCall,
+                                  ExtendedVT);
```
CreateFPExt?

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+      continue;
+    }
+    // There is no intrinsic with his level of precision, truncate the shadow.
+    Args.push_back(
+        Builder.CreateCast(Instruction::FPTrunc, Shadow, IntrinsicArgTy));
```
CreateFPTrunc?

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+  IRBuilder<> Builder(getNextInstructionOrDie(Store));
+  Builder.SetCurrentDebugLocation(Store.getDebugLoc());
+  Value *Dst = PtrOp;
+  const DataLayout &DL =
+      Store.getParent()->getParent()->getParent()->getDataLayout();
```
I'd expect this to be a class member

arsenm commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2236 @@
+  const DataLayout &DL =
+      Store.getParent()->getParent()->getParent()->getDataLayout();
+  TypeSize SlotSize = DL.getTypeStoreSize(VT);
+  assert(!SlotSize.isScalable() && "unsupported");
```
Can you just ignore the scalable operations instead of dying on them?

alexander-shaposhnikov replied:
we do not die on them - we skip them earlier (getExtendedFPType returns nullptr for them now), however, if one day someone gets to adding support for scalable vectors such checks might be of some help.

alexander-shaposhnikov replied:
it's not a problem to remove such checks (if you like), but I kept them for this reason (they might help in the future)

alexander-shaposhnikov commented:
Addressed comments, code cleanup.

vitalybuka commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2128,6 +2128,8 @@ static Attribute::AttrKind getAttrFromCode(uint64_t Code) {
     return Attribute::SanitizeThread;
   case bitc::ATTR_KIND_SANITIZE_MEMORY:
     return Attribute::SanitizeMemory;
+  case bitc::ATTR_KIND_SANITIZE_NUMERICAL_STABILITY:
```
Would be possible to extract attr, bitcode into a separate patch?

alexander-shaposhnikov replied:
Yeah, i can do that.

alexander-shaposhnikov commented:
Addressed comments, rebased the patch. @arsenm, @vitalybuka - happy to do more passes over it, otherwise maybe
we can try landing the patch / getting it through the build bots (this would unblock further progress).

vitalybuka commented on llvm/include/llvm/Transforms/Instrumentation/NumericalStabilitySanitizer.h:
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
```
`///` looks unusual here

nikic commented:
Is there an RFC for this new sanitizer somewhere?

vitalybuka commented on llvm/include/llvm/Transforms/Instrumentation/NumericalStabilitySanitizer.h:
```diff
@@ -0,0 +1,36 @@
+/// functions aren't declared yet, the pass inserts the declarations.
+struct NumericalStabilitySanitizerPass
+    : public PassInfoMixin<NumericalStabilitySanitizerPass> {
+  PreservedAnalyses run(Module &M, ModuleAnalysisManager &AM);
```
Would be nice to pre-commit  tests without instrumentation. So it would be easy to see diff of how intrumentation transform the code.
E.g. this patch with empty ::run() body

vitalybuka replied:
Not strong request, if you see no value, leave as is.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+        " domain (`(trunc(x_shadow) == 0.0f) == (x == 0.0f)`). This helps "
+        "catch the case when `x_shadow` is accurate enough (and therefore "
+        "close enough to zero) so that `trunc(x_shadow)` is zero even though "
+        "both `x` and `x_shadow` are not. "),
```
inconsistent use of `.`
and trailing space in description.

I guess convention is to avoid `.` @MaskRay

alexander-shaposhnikov replied:
Updated.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+
+bool NumericalStabilitySanitizer::sanitizeFunction(
+    Function &F, const TargetLibraryInfo &TLI) {
+  if (!F.hasFnAttribute(Attribute::SanitizeNumericalStability))
```
Maybe
```
if (F.empty())
    return false;
```

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  //    incorrect if a non-instrumented function stores to memory, or if
+  //    memory is stored to through a char pointer.
+  //  - A shadow stack, which holds `s(v)` for any floating-point argument `v`
+  //    of a call to an instrumented function. This allows
```
Any thoughts why shadow stack better/worse then tls approach used by msan?

vitalybuka replied:
cc @eugenis @ramosian-glider

vitalybuka replied:
`__nsan_shadow_ret_val` is like msan, BTW

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+bool NumericalStabilitySanitizer::instrumentMemIntrinsic(MemIntrinsic *MI) {
+  IRBuilder<> Builder(MI);
+  if (MemSetInst *M = dyn_cast<MemSetInst>(MI)) {
+    Builder.SetCurrentDebugLocation(M->getDebugLoc());
```
I believe builder constructed as above will have debug location on that instruction

alexander-shaposhnikov replied:
yeah

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  }
+
+  if (CallBase *CB = dyn_cast<CallBase>(&Inst)) {
+    maybeAddSuffixForNsanInterface(CB);
```
I guess maybeAddSuffixForNsanInterface applies to callsite created by this pass

why not create them with correct parameter in the first place?

alexander-shaposhnikov replied:
no, such callsites originate from the source code, a user can declare "extern void __nsan_dump_shadow_mem(...);" and then call this function to dump the shadow memory. What this pass does - it sets the parameter to the configuration used by the sanitizer (unknown to the user).

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+    Inst.replaceSuccessorWith(NextBB, NewBB);
+
+    IRBuilder<> Builder(NewBB);
+    Builder.SetCurrentDebugLocation(Invoke->getDebugLoc());
```
IRBuilder::IRBuilder() calls SetInsertPoint which calls SetCurrentDebugLocation
So I suspect most calls SetCurrentDebugLocation should be removed

alexander-shaposhnikov replied:
so this is for a newly created basic block, in this case IRBuiilder doesn't really have anything to copy the debug location from.
For the cases where we create IRBuilder from an instruction "Inst", SetInsertPoint (inside the constructor) would set it to the debug loc for "Inst", usually that's not what we want ("Inst" follows the instruction that we actually care about).

alexander-shaposhnikov replied:
Cleaned up the redundant calls of SetCurrentDebugLocation.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  TypeSize SlotSize = DL.getTypeStoreSize(VT);
+  assert(!SlotSize.isScalable() && "unsupported");
+  const auto LoadSizeBytes = SlotSize.getFixedValue();
+  Value *ValueSize = Builder.Insert(Constant::getIntegerValue(
```
It's equivalent to simpler:
```
Value *ValueSize = Constant::getIntegerValue(
      IntptrTy, APInt(IntptrTy->getPrimitiveSizeInBits(), LoadSizeBytes));
```

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+    // value from the source. Note that we cannot use __nsan_copy_values()
+    // here, because that will not work when there is a write to memory in
+    // between the load and the store, e.g. in the case of a swap.
+    Type *ShadowTypeIntTy = Type::getIntNTy(Context, 8 * LoadSizeBytes);
```
There is getTypeStoreSizeInBits

alexander-shaposhnikov replied:
see Section 3.2.3 in https://arxiv.org/pdf/2102.12782, perhaps, ideally this needs to be refactored ...

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  // Split the basic block. On mismatch, we'll jump to the new basic block with
+  // a call to the runtime for error reporting.
+  BasicBlock *FCmpBB = FCmp.getParent();
+  BasicBlock *NextBB = FCmpBB->splitBasicBlock(getNextInstructionOrDie(FCmp));
```
predicting positive branch makes very noticable difference 
`MDBuilder(*C).createUnlikelyBranchWeights()`

vitalybuka replied:
^ applies to CreateCondBr

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  return Shadow;
+}
+
+Value *NumericalStabilitySanitizer::handleExt(FPExtInst &Ext, Type *VT,
```
this and all above could be nicer with IRBuilder

vitalybuka replied:
WDYT?

alexander-shaposhnikov replied:
working on it

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  getIntrinsicInfoTableEntries(WidenedId, Table);
+  SmallVector<Type *, 4> ArgTys;
+  ArrayRef<Intrinsic::IITDescriptor> TableRef = Table;
+  [[maybe_unused]] const Intrinsic::MatchIntrinsicTypesResult Res =
```
```
assert(Intrinsic::matchIntrinsicSignature(WidenedFnTy, TableRef, ArgTys) ==
             Intrinsic::MatchIntrinsicTypes_Match &&
         "invalid widened intrinsic");
```

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+    return Shadow;
+  }
+  if (BinaryOperator *BinOp = dyn_cast<BinaryOperator>(&Inst)) {
+    IRBuilder<> Builder(getNextInstructionOrDie(*BinOp));
```
why it's need to be getNextInstructionOrDie, why not to insert before current one?

alexander-shaposhnikov replied:
So this pattern is used in many places,
i guess the reason is readability, it looks more natural to put them after the original instruction.

vitalybuka replied:
In sanitizers we use as much as possible pre-instrumentation. Also good for forward iterator invalidation.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+  if (LoadInst *Load = dyn_cast<LoadInst>(&Inst)) {
+    return handleLoad(*Load, VT, ExtendedVT);
+  }
+  if (CallInst *Call = dyn_cast<CallInst>(&Inst)) {
```
probably you can create single function level IRBulder here and simplify code.

alexander-shaposhnikov replied:
Done

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2219 @@
+    // All operands have shadows. Create a shadow for the current value.
+    Value *Shadow = createShadowValueWithOperandsAvailable(*I, TLI, Map);
+    Map.setShadow(I, Shadow);
+    DfsStack.pop_back();
```
DFS looks  confusing
can you pop just after copying .back() into local:

```
 while (!DfsStack.empty()) {
    Instruction *I = DfsStack.back();
    DfsStack.pop_back(); 
```

alexander-shaposhnikov replied:
So here we do pop_back when the DFS is going "up" (not "down") 
(so "I" should not be removed from the stack before all the operands are processed)"

vitalybuka commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2205 @@
+  assert(!isa<PHINode>(Root) && "phi nodes should already have shadows");
+
+  std::vector<Instruction *> DfsStack(1, &Root);
+  while (!DfsStack.empty()) {
```
```suggestion
  while (!DfsStack.empty()) {
     // Ensure that all operands to the instruction have shadows before
    // proceeding.
    Instruction *I = DfsStack.back();
    DfsStack.pop_back();

        // The shadow for the instruction might have been created deeper in the DFS,
    // see `forward_use_with_two_uses` test.
    if (Map.hasShadow(I))
      continue;

    bool MissingShadow = false;
    for (Value *Op : I->operands()) {
      Type *VT = Op->getType();
      if (!Config.getExtendedFPType(VT))
        continue; // Not an FT value.
      if (Map.hasShadow(Op))
        continue; // Shadow is already available.
      MissingShadow = true;
      DfsStack.push_back(cast<Instruction>(Op));
    }
    if (MissingShadow)
      continue; // Process operands and come back to this instruction later.

    // All operands have shadows. Create a shadow for the current value.
    Value *Shadow = createShadowValueWithOperandsAvailable(*I, TLI, Map);
    Map.setShadow(I, Shadow);
  }
```

vitalybuka replied:
I see

MaskRay commented:
I'll take a quick look at this patch today, but probably more focused on the style.

For easy reference, the description/commit message should link to the RFC discussion: https://discourse.llvm.org/t/rfc-revival-of-numerical-sanitizer/79601
People have high-level suggestions are recommended to comment there.

**There has been significant offline interest in this revival.**
I have chatted with @alexander-shaposhnikov, who mentioned that several companies and organizations expressing enthusiasm.
I personally mentioned this effort last December to a few folks working on floating-point arithmetic for C library functions,
and they're interested in the feature. While they support the idea, building llvm-project isn't familiar territory for them, and they'd prefer having the code in the tree for further suggestions.

However, **there's room for improvement in the process**. 
Ideally, future RFCs could be posted sooner to generate early discussion. While this revival builds on @legrosbuffle's well-supported earlier work (https://reviews.llvm.org/D97854), the delay highlights the need for a streamlined process for future revivals.
To better gauge industry interest and facilitate collaboration, the discourse post is a place to explicitly record support from companies and organizations in the RFC process.

In light of recent proposals involving instrumentation and compiler-rt, we're refining the RFC process to establish clearer acceptance criteria.

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file is a part of NumericalStabilitySanitizer.
```
This is the main part of the instrumentation and deserves a more detailed comment.

alexander-shaposhnikov replied:
done

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+
+#include "llvm/Transforms/Instrumentation/NumericalStabilitySanitizer.h"
+
+#include <cstdint>
```
system header after llvm headers https://llvm.org/docs/CodingStandards.html#include-style

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+        " domain (`(trunc(x_shadow) == 0.0f) == (x == 0.0f)`). This helps "
+        "catch the case when `x_shadow` is accurate enough (and therefore "
+        "close enough to zero) so that `trunc(x_shadow)` is zero even though "
+        "both `x` and `x_shadow` are not."),
```
omit trailing period

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+// load rather than store.
+// FIXME: provide a way to resume computations from the FT value when the load
+// is inconsistent. This ensures that further computations are not polluted.
+static cl::opt<bool> ClCheckLoads("nsan-check-loads", cl::init(false),
```
remove default `cl::init(false)`

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+static constexpr StringLiteral kNsanInitName("__nsan_init");
+
+// The following values must be kept in sync with the runtime.
+static constexpr int kShadowScale = 2;
```
namespace-scope const variables are automatically internal linkage. `static` is usually omitted

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  }
+
+private:
+  // Must be kept in sync with the runtime.
```
This can be more specific like `nsan/nsan_stats.h`

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+
+private:
+  // Must be kept in sync with the runtime.
+  enum CheckType {
```
How about `enum class CheckType : uint8_t` and moving this after `Address` to make the structure smaller?

alexander-shaposhnikov replied:
they values are passed as int to the runtime, i think keeping it as int is simpler for now,
changing the order of fields is indeed a good idea https://godbolt.org/z/afTTv4YMx

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+
+  // Sets the shadow value for a value. Asserts that the value does not already
+  // have a value.
+  void setShadow(Value *V, Value *Shadow) {
```
Use `&` to assert non-nullness

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  ContinueWithShadow = 0,
+  ResumeFromValue = 1,
+};
+Value *NumericalStabilitySanitizer::emitCheckInternal(Value *V, Value *ShadowV,
```
blank line before this declaration

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+
+  Function *F = FCmp.getFunction();
+  if (CheckFunctionsFilter && !CheckFunctionsFilter->match(F->getName())) {
+    return;
```
remove braces

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+    } else if (FT->isDoubleTy()) {
+      Callee = &(NsanFCmpFail[kDouble]);
+    } else if (FT->isX86_FP80Ty()) {
+      // FIXME: make NsanFCmpFailLongDouble work.
```
Prefer TODO if it's not a bug

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  static const WidenedIntrinsic kWidenedIntrinsics[];
+};
+
+FunctionType *Make_Double_Double(LLVMContext &C) {
```
`functionName`

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  IRBuilder<> Builder(getNextInstructionOrDie(Inst));
+  Builder.SetCurrentDebugLocation(Inst.getDebugLoc());
+
+  if (FPTruncInst *Trunc = dyn_cast<FPTruncInst>(&Inst))
```
`if (auto *... = dyn_cast<...>(...))` here and below

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+
+// Moves fast math flags from the function to individual instructions, and
+// removes the attribute from the function.
+// FIXME: Make this controllable with a flag.
```
Prefer TODO if it's not a bug

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+                                                   IRBuilder<> &Builder) {
+  // We cannot look inside inline asm, just expand the result again.
+  if (Call.isInlineAsm()) {
+    return Builder.CreateFPExt(&Call, ExtendedVT);
```
drop braces

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  ++NumInstrumentedFTCalls;
+  return Shadow;
+
+  // NsanShadowRetTag; or it is and it will always do so.
```
comment after the last return is confusing. Reorganize it?

alexander-shaposhnikov replied:
removed

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  std::vector<Instruction *> OriginalInstructions;
+  for (auto &BB : F) {
+    for (auto &Inst : BB) {
+      OriginalInstructions.emplace_back(&Inst);
```
drop braces

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2191 @@
+  if (MemSetInst *M = dyn_cast<MemSetInst>(MI)) {
+    Builder.CreateCall(
+        NsanSetValueUnknown,
+        {// Address
```
Recommended style: `/*Address=*/XXX`

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2182 @@
+};
+
+const KnownIntrinsic::LFEntry KnownIntrinsic::kLibfuncIntrinsics[] = {
+    {LibFunc_sqrtf, "llvm.sqrt.f32"},           //
```
delete `//`?

MaskRay commented on llvm/lib/Transforms/Instrumentation/NumericalStabilitySanitizer.cpp:
```diff
@@ -0,0 +1,2182 @@
+  Value *ICmpEQ = Builder.CreateICmpEQ(
+      CheckResult,
+      ConstantInt::get(Builder.getInt32Ty(),
+                       static_cast<int>(ContinuationType::ResumeFromValue)));
```
ResumeFromValue is 1. Testing it is slightly slower than testing 0 (ContinueWithShadow)

MaskRay committed:
nsan instrumentation: code style

MaskRay reviewed (approved):
I made a few nits, then I realized that I can just publish https://github.com/MaskRay/llvm-project/tree/nsan The top commit contains code style changes.

alexander-shaposhnikov commented:
Reverted for now, will look into the build bot breakage.
